samples/apps/tinyRA/tinyra/tui.py

```diff
@@ -0,0 +1,1186 @@
+                    await conn.commit()
+                    return id
+    except aiosqlite.Error as e:
+        print(f"Error inserting or updating chat message: {e}")
+
+
+def function_names_to_markdown_table(file_path: str) -> str:
+    """
+    Given a python file, extract the function names and docstrings
+    and return them as a markdown table.
+
+    Args:
+        file_path: the path to the python file
+
+    Returns:
+        A markdown table with function names and docstrings.
+    """
+    with open(file_path, "r") as f:
+        content = f.read()
+    tree = ast.parse(content)
+
+    table = "## Available Functions\n\n"
+    for i, item in enumerate(tree.body):
+        if isinstance(item, ast.FunctionDef):
+            name = item.name
+            docstring = ast.get_docstring(item) or "No docstring available"
+            docstring = docstring.split("\n")[0]
+            table += f"- `{name}`: \n{docstring}\n"
+    return table
+
+
+def get_available_functions() -> str:
+    """
+    Get the available functions from the agent_utils.py file
+    and return them as a markdown table.
+    """
+    UTILS_FILE = os.path.join(APP_CONFIG.get_data_path(), "agent_utils.py")
+    if not os.path.exists(UTILS_FILE):
+        # create the file if it does not exist
+        with open(UTILS_FILE, "w") as f:
+            f.write("")
+
+    markdown_table = function_names_to_markdown_table(UTILS_FILE)
+    return markdown_table
+
+
+def json_to_markdown_code_block(json_data: dict, pretty_print: bool = True) -> str:
+    """
+    Converts a JSON object to a markdown code block.
+
+    Args:
+        json_data (dict): The JSON object to convert.
+        pretty_print (bool, optional): Whether to pretty print the JSON. Defaults to True.
+
+    Returns:
+        str: The markdown code block representing the JSON object.
+    """
+    if pretty_print:
+        json_string = json.dumps(json_data, indent=2)
+    else:
+        json_string = json.dumps(json_data)
+
+    markdown_code_block = f"```json\n{json_string}\n```"
+    return markdown_code_block
+
+
+def message2markdown(message: Dict[str, str]) -> str:
+    """
+    Convert a message to markdown that can be displayed in the chat display.
+
+    Args:
+        message: a message
+
+    Returns:
+        A markdown string.
+    """
+    role = message["role"]
+    if role == "user":
+        display_name = APP_CONFIG.get_user_name()
+    elif role == "assistant" or role == "info":
+        display_name = "TinyRA"
+    else:
+        display_name = "."
+
+    if role == "info":
+        display_id = "\U0001F4AD" * 3
+    else:
+        display_id = message["id"]
+
+    content = message["content"]
+
+    return f"[{display_id}] {display_name}: {content}"
+
+
+class ReactiveMessage(Markdown):
+    """
+    A reactive markdown widget for displaying assistant messages.
+    """
+
+    message = reactive({"role": None, "content": None, "id": None})
```

Comment:
I'd be tempted to use a named tuple here rather than a dict. Textual reactives aren't recursive, and a immutable object would help to show that.